packages/zowe-explorer-api/src/fs/types/abstract.ts

```diff
@@ -0,0 +1,149 @@
+/**
+ * This program and the accompanying materials are made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-v20.html
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Copyright Contributors to the Zowe Project.
+ *
+ */
+
+import { Duplex } from "stream";
+import { IProfileLoaded } from "@zowe/imperative";
+import { IDownloadSingleOptions, IUploadOptions, IZosFilesResponse } from "@zowe/zos-files-for-zowe-sdk";
+import * as vscode from "vscode";
+import { ZosEncoding } from "../../tree";
+
+export enum ZoweScheme {
+    DS = "zowe-ds",
+    Jobs = "zowe-jobs",
+    USS = "zowe-uss",
+}
+
+export enum ConflictViewSelection {
+    UserDismissed = 0,
+    Compare = 1,
+    Overwrite = 2,
+}
+
+export type DeleteMetadata = {
+    entryToDelete: IFileSystemEntry;
+    parent: DirEntry;
+    parentUri: vscode.Uri;
+};
+
+export interface Conflictable {
+    getContents: (filePath: string, options: IDownloadSingleOptions) => Promise<IZosFilesResponse>;
+    uploadFromBuffer?(buffer: Buffer, filePath: string, options?: IUploadOptions): Promise<string | IZosFilesResponse>;
+}
+
+export class BufferBuilder extends Duplex {
+    private chunks: Uint8Array[];
+
+    public constructor() {
+        super();
+        this.chunks = [];
+    }
+
+    public _write(chunk: any, encoding: BufferEncoding, callback: (error?: Error) => void): void {
+        this.chunks.push(chunk);
+        callback();
+    }
+
+    public _read(_size: number): void {
+        const concatBuf = Buffer.concat(this.chunks);
+        this.push(concatBuf);
+        this.push(null);
+    }
+}
+
+export const FS_PROVIDER_DELAY = 5;
+
+export type EntryMetadata = {
+    profile: IProfileLoaded;
+    path: string;
+};
+
+export type ConflictData = {
+    contents: Uint8Array;
+    etag?: string;
+    size: number;
+};
+
+export interface IFileSystemEntry extends vscode.FileStat {
+    name: string;
+    metadata: EntryMetadata;
+    type: vscode.FileType;
+    wasAccessed: boolean;
+    data?: Uint8Array;
+}
+
+export class FileEntry implements IFileSystemEntry {
```

Comment:
Curious why we are implementing the interface vs converting `IFileSystemEntry` into a class and extending it.
If it's possible, I think it will reduce having to create/define the same properties in `FileEntry` (and `DirEntry`)

The reason I ask is because I don't see any abstract methods implemented in either of the classes.

Reply:
I'm hesitant to use a common class over an interface because a common class will introduce a layer of inheritance for the entries in the provider. With an interface, inheritance is avoided and it is just leveraged for type checking.

Another scenario that I thought of is that if we change the base class, we have to update the implementation for all subclasses. With an interface, we can add optional properties in the future that do not have to be defined by the classes that implement it.

Essentially, I think either option (base class or interface) would work here, so I'm willing to reconsider - but all in all, this interface is mostly for adding type checks without the overhead of a base class. If we see a benefit in changing this to a base class though, I'd be happy to do so 😁

Reply:
Thanks for providing the reasoning behind why we went with an interface instead of the possible extra overhead of the base class. I don't have a strong opinion one way or the other since they do seem comparable (in terms of memory footprint and other aspects). That being said, I still like this implementation very much. 😋 

No need to change it to a base class unless others have a strong opinion in favor of it 😋 
Resolving!